### ISSUE TITLE
[Performance] runtime strides in make_block_ptr are 5.4x slower than constexpr strides on Ascend 910B2

### ISSUE THREAD

**Issue by rogerchen838** [triton-lang/triton-ascend#1673]
## Summary

On Ascend 910B2, a regular 2-D strided load expressed with
`tl.make_block_ptr` becomes about **5.4x slower when its strides are runtime
values instead of `tl.constexpr` values**, even though the runtime values are
identical and the block shape, grid, input, output, and computation are the
same.

Splitting the variables shows that runtime `shape` is not the cause:

| block-pointer metadata | median kernel duration |
|---|---:|
| static shape, static strides | 1,056.920 us |
| runtime shape, static strides | 1,059.960 us |
| static shape, runtime strides | 5,769.700 us |
| runtime shape, runtime strides | 5,760.750 us |

The device times come from NPU Profiler
`ASCEND_PROFILER_OUTPUT/kernel_details.csv:Duration(us)`, not host timing.
All four kernels produce the exact same result.

This suggests the Ascend block-pointer lowering has an efficient path for
compile-time strides, but does not generate a comparable runtime-configured
strided transfer when the same strides arrive as scalar kernel arguments.

## Minimal reproducer

This is direct Triton JIT code and does not use `torch.compile` or Inductor.
The logical matrix has shape `[4096, 1024]`.  Consecutive logical columns are
32 FP32 elements apart in physical storage.  Each program processes the same
`[32, 128]` logical tile.

```python
import csv
from pathlib import Path

import torch
import torch_npu
import triton
import triton.language as tl


@triton.jit
def block_ptr_dynamic_kernel(
    storage, out, rows, cols, row_stride, element_stride,
    BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr,
):
    pm, pn = tl.program_id(0), tl.program_id(1)
    src = tl.make_block_ptr(
        base=storage,
        shape=(rows, cols),
        strides=(row_stride, element_stride),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    dst = tl.make_block_ptr(
        base=out,
        shape=(rows, cols),
        strides=(cols, 1),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    x = tl.load(src)
    tl.store(dst, x * 1.001 + 0.125)


@triton.jit
def block_ptr_dynamic_shape_static_stride_kernel(
    storage, out, rows, cols,
    ROW_STRIDE: tl.constexpr, ELEMENT_STRIDE: tl.constexpr,
    BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr,
):
    pm, pn = tl.program_id(0), tl.program_id(1)
    src = tl.make_block_ptr(
        base=storage,
        shape=(rows, cols),
        strides=(ROW_STRIDE, ELEMENT_STRIDE),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    dst = tl.make_block_ptr(
        base=out,
        shape=(rows, cols),
        strides=(cols, 1),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    x = tl.load(src)
    tl.store(dst, x * 1.001 + 0.125)


@triton.jit
def block_ptr_static_shape_dynamic_stride_kernel(
    storage, out, row_stride, element_stride,
    ROWS: tl.constexpr, COLS: tl.constexpr,
    BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr,
):
    pm, pn = tl.program_id(0), tl.program_id(1)
    src = tl.make_block_ptr(
        base=storage,
        shape=(ROWS, COLS),
        strides=(row_stride, element_stride),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    dst = tl.make_block_ptr(
        base=out,
        shape=(ROWS, COLS),
        strides=(COLS, 1),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    x = tl.load(src)
    tl.store(dst, x * 1.001 + 0.125)


@triton.jit
def block_ptr_static_kernel(
    storage, out,
    ROWS: tl.constexpr, COLS: tl.constexpr,
    ROW_STRIDE: tl.constexpr, ELEMENT_STRIDE: tl.constexpr,
    BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr,
):
    pm, pn = tl.program_id(0), tl.program_id(1)
    src = tl.make_block_ptr(
        base=storage,
        shape=(ROWS, COLS),
        strides=(ROW_STRIDE, ELEMENT_STRIDE),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    dst = tl.make_block_ptr(
        base=out,
        shape=(ROWS, COLS),
        strides=(COLS, 1),
        offsets=(pm * BLOCK_M, pn * BLOCK_N),
        block_shape=(BLOCK_M, BLOCK_N),
        order=(1, 0),
    )
    x = tl.load(src)
    tl.store(dst, x * 1.001 + 0.125)


ROWS, COLS, ELEMENT_STRIDE = 4096, 1024, 32
ROW_STRIDE = COLS * ELEMENT_STRIDE
BLOCK_M, BLOCK_N = 32, 128
grid = (ROWS // BLOCK_M, COLS // BLOCK_N)

storage = torch.randn(
    (ROWS, COLS * ELEMENT_STRIDE), device="npu", dtype=torch.float32
)
packed = storage[:, ::ELEMENT_STRIDE].contiguous()
out = torch.empty_like(packed)  # shared output allocation for all variants

calls = [
    ("all dynamic", lambda: block_ptr_dynamic_kernel[grid](
        storage, out, ROWS, COLS, ROW_STRIDE, ELEMENT_STRIDE,
        BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N)),
    ("dynamic shape/static strides",
     lambda: block_ptr_dynamic_shape_static_stride_kernel[grid](
         storage, out, ROWS, COLS,
         ROW_STRIDE=ROW_STRIDE, ELEMENT_STRIDE=ELEMENT_STRIDE,
         BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N)),
    ("static shape/dynamic strides",
     lambda: block_ptr_static_shape_dynamic_stride_kernel[grid](
         storage, out, ROW_STRIDE, ELEMENT_STRIDE,
         ROWS=ROWS, COLS=COLS, BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N)),
    ("all static", lambda: block_ptr_static_kernel[grid](
        storage, out, ROWS=ROWS, COLS=COLS,
        ROW_STRIDE=ROW_STRIDE, ELEMENT_STRIDE=ELEMENT_STRIDE,
        BLOCK_M=BLOCK_M, BLOCK_N=BLOCK_N)),
]

reference = packed * 1.001 + 0.125
for name, call in calls:
    call()
    torch.npu.synchronize()
    torch.testing.assert_close(out, reference, rtol=0, atol=0)
    print("correctness: PASS", name)

# Compile/device warmup outside the recorded profiler interval.
for _ in range(5):
    for _, call in calls:
        call()
torch.npu.synchronize()

samples = 12
profile_dir = "block_ptr_runtime_stride_profile"
cfg = torch_npu.profiler._ExperimentalConfig(
    aic_metrics=torch_npu.profiler.AiCMetrics.PipeUtilization,
    profiler_level=torch_npu.profiler.ProfilerLevel.Level1,
    l2_cache=False,
    data_simplification=False,
)
with torch_npu.profiler.profile(
    activities=[torch_npu.profiler.ProfilerActivity.NPU],
    schedule=torch_npu.profiler.schedule(
        wait=0, warmup=1, active=samples * len(calls), repeat=1
    ),
    on_trace_ready=torch_npu.profiler.tensorboard_trace_handler(profile_dir),
    experimental_config=cfg,
) as prof:
    # Profiler state-machine warmup; not recorded.
    calls[0][1](); torch.npu.synchronize(); prof.step()
    for sample in range(samples):
        start = sample % len(calls)
        for _, call in calls[start:] + calls[:start]:
            call(); torch.npu.synchronize(); prof.step()

csv_path = next(Path(profile_dir).rglob("kernel_details.csv"))
kernel_names = [
    "block_ptr_dynamic_kernel",
    "block_ptr_dynamic_shape_static_stride_kernel",
    "block_ptr_static_shape_dynamic_stride_kernel",
    "block_ptr_static_kernel",
]
durations = {name: [] for name in kernel_names}
with csv_path.open(newline="") as f:
    for row in csv.DictReader(f):
        for name in kernel_names:
            if name in row["Name"]:
                durations[name].append(float(row["Duration(us)"]))
for name, values in durations.items():
    assert len(values) == samples, (name, len(values))
    values.sort()
    print(name, "median(us)=", (values[5] + values[6]) / 2)
```

I profiled the calls after compilation and device warmup.  The profiler
schedule records 12 launches of each kernel.  The first case is rotated once
per sampling round so no variant gets a fixed execution-order advantage.  I
then select each kernel by its unique name in `kernel_details.csv` and compute
p20 / median / p80 over `Duration(us)`.

## Results

Shape `(4096, 1024)`, physical inner stride 32, tile `(32, 128)`, FP32:

| variant | p20 | median | p80 | versus all-static |
|---|---:|---:|---:|---:|
| runtime shape, runtime strides | 5,745.200 us | 5,760.750 us | 5,773.820 us | 5.450x |
| runtime shape, static strides | 1,057.520 us | 1,059.960 us | 1,061.120 us | 1.003x |
| static shape, runtime strides | 5,762.500 us | 5,769.700 us | 5,775.752 us | 5.459x |
| static shape, static strides | 1,055.424 us | 1,056.920 us | 1,060.888 us | 1.000x |

An additional run with physical inner stride 16 showed the same trend:

| metadata | median |
|---|---:|
| runtime shape and strides | 6,047.970 us |
| static shape and strides | 1,393.350 us |
| dynamic/static ratio | 4.340x |

For comparison, explicit pointer arithmetic follows almost exactly the same
paths as `make_block_ptr`:

| addressing | runtime metadata | static metadata |
|---|---:|---:|
| `make_block_ptr` | 5,760.750 us | 1,056.920 us |
| explicit pointer offsets | 5,774.180 us | 1,058.690 us |

Thus the block pointer's structured layout information currently provides no
performance advantage over explicit offsets when strides are runtime values.

A packed-input kernel was about 45.6 us in the same profiling run.  This is
only a hardware upper-bound reference: the packed kernel has a different
physical layout and the cost of packing is intentionally not included.  It is
not used as the primary baseline for the reported 5.4x regression.

## Expected behavior / request

`tl.make_block_ptr` semantically supports runtime shape and stride operands.
It would be useful for the Ascend lowering to retain the efficient structured
strided transfer when the stride is supplied at runtime, for example through a
runtime-configured strided DMA or another suitable path, instead of falling
back to the much slower behavior seen above.

At minimum, could the expected performance support and current limitations
for runtime block-pointer strides on Ascend be documented?

Related: #434 discussed dynamic/non-power-of-two structured strides for the
Ascend 910_95 SIMT path and is now closed as stale.  This report is a direct
910B2 Triton JIT reproducer and isolates runtime strides from runtime shapes.

## Environment

- Hardware: 8 x Huawei Ascend 910B2; this benchmark used NPU 0 only
  - Product: `IT21HMDC_Bin0`
  - One chip per card, 24 AICores per chip
  - AICore nominal frequency reported by `npu-smi`: 1800 MHz
  - HBM reported per card: 65,536 MiB
- Driver / board software (`npu-smi`): `26.0.rc1`
- Firmware: `9.0.0.0.205`
- CANN toolkit and operator packages: `9.1.0-beta.3`
- Architecture: AArch64
- Container OS: Ubuntu 22.04.5 LTS
- Host kernel visible in the container:
  `6.6.0-132.0.0.111.oe2403sp3.aarch64`
- glibc: 2.35
- Python: 3.10.20 (conda environment)
- PyTorch: `2.13.0+cpu`
- torch-npu: `2.13.0+git4d712d1`
- Triton-Ascend branch: `release/3.2.2-dev`

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.

**Comment by melo882** [comment]
block_ptr_dynamic_kernel median(us)= 5833.456
block_ptr_dynamic_shape_static_stride_kernel median(us)= 1068.6215
block_ptr_static_shape_dynamic_stride_kernel median(us)= 5835.4065
block_ptr_static_kernel median(us)= 1067.2510000000002

**Comment by rogerchen838** [comment]
Thanks for confirming this. Your numbers reproduce the same split very closely: making the shape dynamic has essentially no effect, while either variant with runtime strides is about 5.5x slower. That narrows the regression to the dynamic-stride lowering path rather than block-pointer shape handling.
